cli/cmd/edges.go

```diff
@@ -0,0 +1,360 @@
+			}
+			if len(msg) > maxMsgLength {
+				maxMsgLength = len(msg)
+			}
+
+		}
+	}
+	switch options.outputFormat {
+	case tableOutput:
+		if len(edgeTables) == 0 {
+			fmt.Fprintln(os.Stderr, "No edges found.")
+			os.Exit(0)
+		}
+		printEdgeTable(edgeTables, w, maxSrcLength, maxDstLength, maxClientLength, maxServerLength, maxMsgLength)
+	case jsonOutput:
+		printEdgesJSON(edgeTables, w)
+	}
+}
+
+func printEdgeTable(edges map[string]*edgeRow, w *tabwriter.Writer, maxSrcLength, maxDstLength, maxClientLength, maxServerLength, maxMsgLength int) {
+	srcTemplate := fmt.Sprintf("%%-%ds", maxSrcLength)
+	dstTemplate := fmt.Sprintf("%%-%ds", maxDstLength)
+	clientTemplate := fmt.Sprintf("%%-%ds", maxClientLength)
+	serverTemplate := fmt.Sprintf("%%-%ds", maxServerLength)
+	msgTemplate := fmt.Sprintf("%%-%ds", maxMsgLength)
+
+	headers := []string{
+		fmt.Sprintf(srcTemplate, srcHeader),
+		fmt.Sprintf(dstTemplate, dstHeader),
+		fmt.Sprintf(clientTemplate, clientHeader),
+		fmt.Sprintf(serverTemplate, serverHeader),
+		fmt.Sprintf(msgTemplate, msgHeader),
+	}
+
+	headers[len(headers)-1] = headers[len(headers)-1] + "\t" // trailing \t is required to format last column
+
+	fmt.Fprintln(w, strings.Join(headers, "\t"))
+
+	sortedKeys := sortEdgesKeys(edges)
+	for _, key := range sortedKeys {
+		values := make([]interface{}, 0)
+		templateString := fmt.Sprintf("%s\t%s\t%s\t%s\t%s\t\n", srcTemplate, dstTemplate, clientTemplate, serverTemplate, msgTemplate)
+
+		values = append(values, []interface{}{
+			edges[key].src,
+			edges[key].dst,
+			edges[key].client,
+			edges[key].server,
+			edges[key].msg,
+		}...)
+
+		fmt.Fprintf(w, templateString, values...)
+
+	}
+}
+
+func renderEdges(buffer bytes.Buffer, options *edgesOptions) string {
+	var out string
+	switch options.outputFormat {
+	case jsonOutput:
+		out = buffer.String()
+	default:
+		// strip left padding on the first column
+		out = string(buffer.Bytes()[padding:])
+		out = strings.Replace(out, "\n"+strings.Repeat(" ", padding), "\n", -1)
+	}
+
+	return out
+}
+
+func sortEdgesKeys(stats map[string]*edgeRow) []string {
+	var sortedKeys []string
+	for key := range stats {
+		sortedKeys = append(sortedKeys, key)
+	}
+	sort.Strings(sortedKeys)
+	return sortedKeys
+}
+
+type edgesJSONStats struct {
+	Src    string `json:"src"`
+	Dst    string `json:"dst"`
+	Client string `json:"client_id"`
+	Server string `json:"server_id"`
+	Msg    string `json:"no_tls_reason"`
+}
+
+func printEdgesJSON(edgeTables map[string]*edgeRow, w *tabwriter.Writer) {
+	// avoid nil initialization so that if there are not stats it gets marshalled as an empty array vs null
+	entries := []*edgesJSONStats{}
+
+	for _, row := range edgeTables {
+		entry := &edgesJSONStats{
+			Src:    row.src,
+			Dst:    row.dst,
+			Client: row.client,
+			Server: row.server,
+			Msg:    row.msg}
+		entries = append(entries, entry)
+	}
+
+	b, err := json.MarshalIndent(entries, "", "  ")
+	if err != nil {
+		log.Error(err.Error())
```

Comment:
i realize we're not consistent about printing errors in the cli (`log.Error` vs. `fmt.Print(os.Stderr...`), but i'd prefer the latter, as logging is more of a server thing, and being explicit about stdout vs. stderr in the cli is a good thing:
```go
fmt.Fprintf(os.Stderr, "Error marshalling JSON: %s", err)
```
(@klingerf and i had a good exchange in some long ago PR i can't find, where he convinced me of this position)

Reply:
Makes sense 👍